Move _thnn_conv2d resize and zero code from codegen to native code.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37958 Kill resize-ing and zero-ing from codegen.
* #37957 Move resize / zero logic for _thnn_conv_depthwise2d from codegen to native code.
* **#37956 Move _thnn_conv2d resize and zero code from codegen to native code.**
* #37955 Move resize logic for bmm from codegen to native code.
* #37907 [RESUBMIT] Kill broadcasting from the codegen layer.

This is basically just doing what the CPU code already does, but with keeping the kernel in THC, unlike in CPU where that has already moved to native.

Differential Revision: [D21433211](https://our.internmc.facebook.com/intern/diff/D21433211)

gchanan committed:
Move _thnn_conv2d resize and zero code from codegen to native code.

This is basically just doing what the CPU code already does, but with keeping the kernel in THC, unlike in CPU where that has already moved to native.

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/nn_parse.py:
```diff
@@ -295,10 +295,10 @@ def initialize_output_arg(arg):
 
         # grad_weight and grad_bias need to be resized and zeroed
-        if arg['name'] == 'grad_weight':
+        if arg['name'] == 'grad_weight' and base['name'] != '_thnn_conv2d':
```
If there's gonna be more of these, maybe it's worth like `PORTED_RESIZE_AND_ZERO = [...]` that gets checked against? 

Actually maybe worth it already given the multiple checks

gchanan replied:
there are only two of them, so I didn't think it was worth it.

bhosmer commented on aten/src/ATen/native/cuda/ConvolutionMM2d.cu:
```diff
@@ -0,0 +1,67 @@
+                                               grad_output, self, weight,
+                                               kernel_size, stride, padding,
+                                               finput, fgrad_input);
+}
```
Are these just a straight transcription of the generated code or is there more to think about here

gchanan replied:
it's not quite a straight transcription, but it is the same as what was already done on the CPU side.  The two differences are:
- the generated non-out variant doesn't call the out variant.  This allows codegen to do some stuff specifically for the out variant but isn't necessary in this case.
- The creation of the outputs in the non-out variants are done in a way that avoids dispatch but is so verbose that no one would actually write it:
```
    auto grad_bias_ = output_mask[2] ? c10::make_intrusive<TensorImpl, UndefinedTensorImpl>(c10::Storage(scalarTypeToTypeMeta(dispatch_scalar_type), 0, allocator(), true),DispatchKey::CUDA).releas\
e() : nullptr;
    auto grad_bias = Tensor(c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl>::reclaim(grad_bias_ == nullptr ? (TensorImpl*)UndefinedTensorImpl::singleton() : (TensorImpl*)grad_bias_));
```

gchanan committed:
Update on "Move _thnn_conv2d resize and zero code from codegen to native code."


This is basically just doing what the CPU code already does, but with keeping the kernel in THC, unlike in CPU where that has already moved to native.

Differential Revision: [D21433211](https://our.internmc.facebook.com/intern/diff/D21433211)

[ghstack-poisoned]

gchanan committed:
Update on "Move _thnn_conv2d resize and zero code from codegen to native code."


This is basically just doing what the CPU code already does, but with keeping the kernel in THC, unlike in CPU where that has already moved to native.

Differential Revision: [D21433211](https://our.internmc.facebook.com/intern/diff/D21433211)

[ghstack-poisoned]
